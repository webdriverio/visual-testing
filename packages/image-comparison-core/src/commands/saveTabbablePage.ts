import type { ScreenshotOutput } from '../helpers/afterScreenshot.interfaces.js'
import drawTabbableOnCanvas from '../clientSideScripts/drawTabbableOnCanvas.js'
import saveFullPageScreen from './saveFullPageScreen.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import type { InternalSaveTabbablePageMethodOptions } from './save.interfaces.js'

/**
 * Saves an image of all tab executions
 */
export default async function saveTabbablePage(
    {
        browserInstance,
        instanceData,
        isNativeContext = false,
        folders,
        tag,
        saveTabbableOptions,
    }: InternalSaveTabbablePageMethodOptions
): Promise<ScreenshotOutput> {
    // 1a. Check if the method is supported in native context
    if (isNativeContext) {
        throw new Error('The method saveTabbablePage is not supported in native context for native mobile apps!')
    }

    // 1b. Inject drawing the tabbables
    await browserInstance.execute(drawTabbableOnCanvas, saveTabbableOptions.wic.tabbableOptions)

    // 2. Create the screenshot
    const fullPageData = await saveFullPageScreen({ browserInstance, folders, instanceData, isNativeContext, saveFullPageOptions: saveTabbableOptions, tag })

    // 3. Remove the canvas
    await browserInstance.execute(removeElementFromDom, 'wic-tabbable-canvas')

    // 4. Return the data
    return fullPageData
}
