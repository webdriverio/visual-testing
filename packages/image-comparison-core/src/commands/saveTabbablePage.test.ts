import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest'
import saveTabbablePage from './saveTabbablePage.js'
import type { InternalSaveTabbablePageMethodOptions } from './save.interfaces.js'
import {
    BASE_CHECK_OPTIONS,
    createMethodOptions
} from '../mocks/mocks.js'

vi.mock('./saveFullPageScreen.js', () => ({
    default: vi.fn().mockResolvedValue({
        devicePixelRatio: 2,
        fileName: 'test-tabbable-page.png'
    })
}))
vi.mock('../clientSideScripts/drawTabbableOnCanvas.js', () => ({
    default: vi.fn()
}))
vi.mock('../clientSideScripts/removeElementFromDom.js', () => ({
    default: vi.fn()
}))

describe('saveTabbablePage', () => {
    let saveFullPageScreen: any
    const executeMock = vi.fn().mockResolvedValue(undefined)

    const baseOptions: InternalSaveTabbablePageMethodOptions = {
        browserInstance: {
            execute: executeMock,
            isAndroid: false,
            isMobile: false
        } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        isNativeContext: false,
        saveTabbableOptions: {
            wic: {
                ...BASE_CHECK_OPTIONS.wic,
                tabbableOptions: {
                    circle: {
                        backgroundColor: 'red',
                        borderColor: 'blue',
                        borderWidth: 2,
                        fontColor: 'white',
                        fontFamily: 'Arial',
                        fontSize: 10,
                        size: 10,
                        showNumber: true
                    },
                    line: {
                        color: 'blue',
                        width: 1
                    }
                }
            },
            method: createMethodOptions()
        },
        tag: 'test-tabbable-page'
    }

    beforeEach(async () => {
        saveFullPageScreen = (await vi.importMock('./saveFullPageScreen.js')).default
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should throw an error when in native context', async () => {
        const options = {
            ...baseOptions,
            isNativeContext: true
        }

        await expect(saveTabbablePage(options)).rejects.toThrow('The method saveTabbablePage is not supported in native context for native mobile apps!')
        expect(executeMock).not.toHaveBeenCalled()
        expect(saveFullPageScreen).not.toHaveBeenCalled()
    })

    it('should save a tabbable page screenshot', async () => {
        const result = await saveTabbablePage(baseOptions)

        expect(result).toMatchSnapshot()
        expect(executeMock).toHaveBeenCalledTimes(2)
        expect(executeMock.mock.calls).toMatchSnapshot()
        expect(saveFullPageScreen.mock.calls).toMatchSnapshot()
    })
})
