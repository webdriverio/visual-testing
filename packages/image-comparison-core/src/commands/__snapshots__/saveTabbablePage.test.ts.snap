// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`saveTabbablePage > should save a tabbable page screenshot 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-tabbable-page.png",
}
`;

exports[`saveTabbablePage > should save a tabbable page screenshot 2`] = `
[
  [
    [MockFunction spy],
    {
      "circle": {
        "backgroundColor": "red",
        "borderColor": "blue",
        "borderWidth": 2,
        "fontColor": "white",
        "fontFamily": "Arial",
        "fontSize": 10,
        "showNumber": true,
        "size": 10,
      },
      "line": {
        "color": "blue",
        "width": 1,
      },
    },
  ],
  [
    [MockFunction spy],
    "wic-tabbable-canvas",
  ],
]
`;

exports[`saveTabbablePage > should save a tabbable page screenshot 3`] = `
[
  [
    {
      "browserInstance": {
        "execute": [MockFunction spy] {
          "calls": [
            [
              [MockFunction spy],
              {
                "circle": {
                  "backgroundColor": "red",
                  "borderColor": "blue",
                  "borderWidth": 2,
                  "fontColor": "white",
                  "fontFamily": "Arial",
                  "fontSize": 10,
                  "showNumber": true,
                  "size": 10,
                },
                "line": {
                  "color": "blue",
                  "width": 1,
                },
              },
            ],
            [
              [MockFunction spy],
              "wic-tabbable-canvas",
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": Promise {},
            },
            {
              "type": "return",
              "value": Promise {},
            },
          ],
        },
        "isAndroid": false,
        "isMobile": false,
      },
      "folders": {
        "actualFolder": "/test/actual",
        "baselineFolder": "/test/baseline",
        "diffFolder": "/test/diff",
      },
      "instanceData": {
        "appName": "TestApp",
        "browserName": "Chrome",
        "browserVersion": "118.0.0.0",
        "deviceName": "iPhone 14",
        "devicePixelRatio": 2,
        "deviceRectangles": {
          "bottomBar": {
            "height": 0,
            "width": 390,
            "x": 0,
            "y": 800,
          },
          "homeBar": {
            "height": 34,
            "width": 390,
            "x": 0,
            "y": 780,
          },
          "leftSidePadding": {
            "height": 0,
            "width": 0,
            "x": 0,
            "y": 0,
          },
          "rightSidePadding": {
            "height": 0,
            "width": 0,
            "x": 0,
            "y": 0,
          },
          "screenSize": {
            "height": 844,
            "width": 390,
          },
          "statusBar": {
            "height": 47,
            "width": 390,
            "x": 0,
            "y": 0,
          },
          "statusBarAndAddressBar": {
            "height": 47,
            "width": 390,
            "x": 0,
            "y": 0,
          },
          "viewport": {
            "height": 733,
            "width": 390,
            "x": 0,
            "y": 47,
          },
        },
        "initialDevicePixelRatio": 2,
        "isAndroid": false,
        "isIOS": true,
        "isMobile": true,
        "logName": "test-log",
        "name": "test-device",
        "nativeWebScreenshot": false,
        "platformName": "iOS",
        "platformVersion": "17.0",
      },
      "isNativeContext": false,
      "saveFullPageOptions": {
        "method": {
          "disableBlinkingCursor": false,
          "disableCSSAnimation": false,
          "enableLayoutTesting": false,
          "enableLegacyScreenshotMethod": false,
          "hideElements": [],
          "hideScrollBars": true,
          "removeElements": [],
          "waitForFontsLoaded": true,
        },
        "wic": {
          "addIOSBezelCorners": false,
          "addressBarShadowPadding": 6,
          "autoElementScroll": true,
          "autoSaveBaseline": false,
          "clearFolder": false,
          "compareOptions": {
            "blockOutSideBar": false,
            "blockOutStatusBar": false,
            "blockOutToolBar": false,
            "createJsonReportFiles": false,
            "diffPixelBoundingBoxProximity": 5,
            "ignoreAlpha": false,
            "ignoreAntialiasing": false,
            "ignoreColors": false,
            "ignoreLess": false,
            "ignoreNothing": false,
            "rawMisMatchPercentage": false,
            "returnAllCompareData": false,
            "saveAboveTolerance": 0,
            "scaleImagesToSameSize": false,
          },
          "disableBlinkingCursor": false,
          "disableCSSAnimation": false,
          "enableLayoutTesting": false,
          "enableLegacyScreenshotMethod": false,
          "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
          "fullPageScrollTimeout": 1500,
          "hideScrollBars": true,
          "isHybridApp": false,
          "savePerInstance": true,
          "tabbableOptions": {
            "circle": {
              "backgroundColor": "red",
              "borderColor": "blue",
              "borderWidth": 2,
              "fontColor": "white",
              "fontFamily": "Arial",
              "fontSize": 10,
              "showNumber": true,
              "size": 10,
            },
            "line": {
              "color": "blue",
              "width": 1,
            },
          },
          "toolBarShadowPadding": 6,
          "userBasedFullPageScreenshot": false,
          "waitForFontsLoaded": true,
        },
      },
      "tag": "test-tabbable-page",
    },
  ],
]
`;
